### ISSUE TITLE
CAP chghost & self

### ISSUE THREAD

**Issue by syzop** [ircv3/ircv3-specifications#324]
I just added ```CAP chghost``` support in UnrealIRCd which made me wonder about the following:
Would it be acceptable, if person X's  username or hostname changes, for a server to also send the CHGHOST message to person X itself? The [specification](http://ircv3.net/specs/extensions/chghost-3.2.html) isn't clear on this.

For client coders this would provide easy notification of user/host changes of not only others but also themselves. For self host changes there is also numeric 396, but there isn't one for username changes. Now, I don't need or want one either, if using "self"  CHGHOST notification would be permitted. It would provide a simple and singular way to notify the client of a user/host change.

The only downside I can think of is that clients may not expect it for source == self. I highly doubt it will cause any issues, though.

Opinions?

**Comment by DanielOaks** [comment]
fwiw I think this makes sense to do and allow

**Comment by jwheare** [comment]
Works for me. Anyone care to PR a suggested wording change and errata?

**Comment by jwheare** [comment]
https://github.com/unrealircd/unrealircd/commit/b7b8e41ac471d700be7d49255a40866dba374c88

For reference, this change was made in Elemental-IRCd too https://github.com/Elemental-IRCd/elemental-ircd/commit/36f09896a19864ad9c6839e628827ea55e7568ab in October 2015 after the initial implementation in September (https://github.com/Elemental-IRCd/elemental-ircd/commit/e84f541ac893daa818a9e8a92034fb824f1e5a88) and they were (I believe) the original proposers of the cap in #17 in 2013. So this was perhaps the original intent. Maybe @Xe or @ariscop can confirm.

Hybrid followed with the same in November 2015 https://github.com/ircd-hybrid/ircd-hybrid/commit/7bc493be91f0338e0c9f9658f540446b49f0027b 

Hybrid/Elemental both use `sendto_common_channels_local` which includes the user.

Charybdis did not include this change in their implementation in November 2015 https://github.com/charybdis-ircd/charybdis/commit/4a5330bb4d2a04c238767213d5621e46d468c7bd (`sendto_common_channels_local_butone` excludes the user) and nor did InspIRCd in December 2015 https://github.com/inspircd/inspircd/commit/ecca6c2300056f1cff1a9c7cac03939ea3f6f9cc#diff-a786522c317191b607a58037bedff213 (`WriteNeighboursWithExt` excludes the user)

In conclusion, this behaviour exists in the wild presumably without issue, but isn't implemented consistently. So, it's safe and the clarification is warranted.

**Comment by syzop** [comment]
Great, thanks for the research! (Oh and I'm not volunteering for the errata :D)

**Comment by syzop** [comment]
Wait. Are you sure you are correct? Normally the sender is skipped in sendto common channels
The hybrid link you gave for example has a 'continue' if 'target_p == user' so I don't think it sends it to the affected user (self).

Also, we at UnrealIRCd will always to the victim/affected user (well, if they have CAP chghost). So even if the user is in no channels.

**Comment by jwheare** [comment]
At least in Elemental, judging from that first linked change, `sendto_common_channels_local_butone` excludes the user and `sendto_common_channels_local` doesn't. Or at least it seems that way. But yes, it looks like Hybrid has just one function with a `touser` bool, and it's in fact *not* set for chghost.

**Comment by syzop** [comment]
Ok, I checked some more. So just to summarize:
Elemental: sends it to the affected user, even if not in any channels
UnrealIRCd: same (from version 4.0.16 onward)
Most others (unchecked, but I think you are quite correct): don't send to affected

I think the errata you suggested would be nice and it should (IMO) also mention that it's okay to send it to the user if it's not in any channels. Or, you can ignore that whole sentence altogether and just say something like "the CHGHOST message MAY also be sent to the affected user" (or SHOULD if it has always been intended).

**Comment by ariscop** [comment]
Not sure what the purpose of excluding the source is, would clients with chghost be sent the numeric instead?

It's essentially nick but for user@host, and notifies the source in the same way

**Comment by ariscop** [comment]
Was discussing the changes with @Xe at the time so it would seem to be the original intent to inform the client

**Comment by jwheare** [comment]
@ariscop thanks for the extra context. I believe `396 RPL_VISIBLEHOST/RPL_HOSTHIDDEN/RPL_YOURDISPLAYEDHOST` gets sent to the user in all cases as well, but as mentioned, doesn't cover any potential change to ident.

**Comment by jwheare** [comment]
Might also be worth mentioning explicitly in the spec that 396 is still expected to be sent. Searching through my channel logs pulled up this question that never got discussed/answered:

```
[2015-11-14 09:34:37] <grawity> should a client receive CHGHOST about themselves?
[2015-11-14 09:34:56] <grawity> and should RPL_VISIBLEHOST be sent if the client enables CHGHOST?
```

**Comment by jwheare** [comment]
Some more log spelunking turned up this relevant discussion about the behaviour of various notify specs/implementations with regards to other numerics being sent and whether the client themselves should be notified.

```
[2016-01-22 18:12:11] <jwheare> should 341 RPL_INVITING be suppressed if invite-notify is enabled?
[2016-01-22 18:12:28] <jwheare> if not you get two notifications when you invite someone
[2016-01-22 18:12:28] <Attila> no
[2016-01-22 18:13:01] <Adam> I dont think the creator of the invite receives the INVITE
    notification, at least not on the impl I wrote for inspircd
[2016-01-22 18:13:24] <Attila> it does on master
[2016-01-22 18:14:28] <jwheare> :madoka.testificate.com 341 jwheare_local jwheare2 #jwheare
[2016-01-22 18:14:28] <jwheare> :jwheare_local!sid61@testificate-cpp.6vf.u5dm0c.IP INVITE jwheare2 #jwheare
[2016-01-22 18:14:55] <jwheare> so INVITE should be suppressed for your own invites?
[2016-01-22 18:15:32] <jwheare> it provides more source information, but you'll already have
    that if it's you
[2016-01-22 18:15:40] <Attila> inspircd already had invite notifications before invite-notify and
    those were sent to all users on the channel
[2016-01-22 18:15:50] <jwheare> as a notice?
[2016-01-22 18:15:52] <Attila> yes
[2016-01-22 18:16:00] <Attila> invite-notify is a logical extension to that
[2016-01-22 18:16:07] <Attila> instead of the notice you get the INVITE if you have invite-notify
[2016-01-22 18:16:09] <grawity> hmm are ACCOUNT events sent to self?
[2016-01-22 18:16:20] <grawity> if they are, and if this isn't yet in the spec, I'd vote to
    suppress 341
[2016-01-22 18:17:39] <Attila> (well, not all users on the channel its configurable whether only
    people who can invite get it, or only ops)
[2016-01-22 18:18:03] <Attila> but the point is the invite-notify INVITE replaces that notice
[2016-01-22 18:18:23] <Attila> i support clearing up this ambiguity from all relevant specs
[2016-01-22 18:28:24] <Attila> s/from/in/
[2016-01-22 18:28:42] <Attila> account-notify away-notify is prob relevant
...
[2016-01-23 01:10:18] <jwheare> grawity, Attila: point of comparison re: self invite-notify with
    other specs. with chghost on insp you don't get a chghost for yourself when toggling
    e.g. +x just the usual 396 RPL_HOSTHIDDEN which appears without the chghost CAP.
```

**Comment by jwheare** [comment]
Also rediscovered this suggested clarification from @RyanSquared that never made it to a PR https://github.com/ircv3/ircv3-specifications/compare/master...RyanSquared:clarify-chghost-3.2.md

**Comment by RyanSquared** [comment]
Sorry, I've kinda been away for a bit, I'll open up a PR if anyone wants?

**Comment by ariscop** [comment]
Looks like this is why i made the change https://github.com/Elemental-IRCd/elemental-ircd/blob/elemental-ircd-6.6.4/tests/chghost.tcl
Should be fairly easy to run the test against any ircd but the test does use chghost from an oper account, i think that's elemental-specific

**Comment by ariscop** [comment]
https://github.com/grawity/irc-docs/blob/master/server/ts6.txt#L293-L299 CHGHOST in TS6 for reference

**Comment by jesopo** [comment]
This is being worked on over at #364

**Comment by kylef** [comment]
I think this issue is resolved, as far as I can tell from [the spec](https://ircv3.net/specs/extensions/chghost-3.2.html):

> Servers SHOULD also send the CHGHOST message to the client whose own username or host changed, if that client also supports the chghost capability.

With errata item:

> Previous versions of this specification did not specify whether or not the client whose own user or host changed should receive the message.

Should we close the issue?
